src/passes/transform/cse.cpp

```diff
@@ -0,0 +1,174 @@
+#include "coreir/passes/transform/cse.h"
+
+using namespace std;
+using namespace CoreIR;
+
+// Desired CoreIR iterators/functional methods
+// breadth-first tree iterator on wireables
+// depth-first tree iterator on wireables
+// leaf-iterator on wireables (Only single bits)
+// breadth-first iteration on types
+// depth-first iterator on types
+// leaf iterator on types 
+// 'All' operator with a predicate
+// 'Any' operator with a predicate
+// 'Map' operator
+// 'Filter' operator (both a modify and a return new)
+
+
+//Some possible (untested) implementations of some of the functions
+template <typename Cont, typename Pred>
+Cont filterOver(const Cont &cont, Pred predicate) {
+  Cont result;
+  std::copy_if(cont.begin(), cont.end(), std::back_inserter(result), predicate);
+  return result
+}
+
+template <typename InCont, typename OutCont, typename Fun>
+OutCont mapOver(const InCont &cont, Fun fun) {
+  OutCont result;
+  std::transform(cont.begin(), cont.end(), std::back_inserter(result), fun);
+  return result
+}
+
+template <typename Cont, typename Pred>
+bool anyOf(const Cont &cont, Pred predicate) {
+  return std::any_of(cont.begin(), cont.end(), predicate);
+}
+
+template <typename Cont, typename Pred>
+bool allOf(const Cont &cont, Pred predicate) {
+  return std::all_of(cont.begin(), cont.end(), predicate);
+}
+
+
+//This is a pass that will do generalized common subexpression elimination.
+//a 'cse' is any time there are two instances of the same module which 
+//are driven by identical values.
+namespace CoreIR {
+
+  namespace Passes {
+
+    string Passes::CSE::ID = "cse";
+    bool CSE::runOnModule(Module* m) {
+      if (!m->hasDef()) {
+        return false;
+      }
+
+      //create map from instance kind to list of instances
+      std::map<Module*, std::vector<Instance*>> modules2Instances;
+      for (auto ipair: m->getInstances()) {
```

Comment:
a lot of these names are not descriptive: `ipair`, `imod`, `iname`. (Again maybe it's understandable to the experts, but e.g. I couldn't really figure out what was going on. Code/names as documentation is always a good idea.)